src/core/LanguageController.ts

```diff
@@ -224,17 +222,270 @@ export default class LanguageController {
     async languageByRef(ref: LanguageRef): Promise<Language> {
         const address = languageAliases[ref.address] ? languageAliases[ref.address] : ref.address
         const language = this.#languages.get(address)
+        //If the language is already installed then just return it
         if(language) {
             return language
         } else {
             let languageMeta = await this.getLanguageExpression(address);
-            if(languageMeta) {
-                const lang = await this.installLanguage(address, languageMeta)
-                return lang!
-            } else {
+            if(!languageMeta) {
                 throw new Error("Language not found by reference: " + JSON.stringify(ref))
             }
+            if (languageMeta.proof.valid != true) {
+                throw new Error("Language to be installed does not have valid proof");
+            }
+            const languageMetaData = languageMeta.data;
+            const languageAuthor = languageMeta.author;
+            //@ts-ignore
+            const trustedAgents: string[] = this.#context.agent.getTrustedAgents();
+            //Check if the author of the language is in the trusted agent list the current agent holds, if so then go ahead and install
+            if (trustedAgents.find((agent) => agent === languageAuthor)) {
+                //Get the language source so we can generate a hash and check against the hash given in the language meta information
+                const languageSource = await this.getLanguageSource(address);
+                if (!languageSource) {
+                    throw new Error("Could not get languageSource for language")
+                }
+                const languageHash = await this.ipfsHash(languageSource);
+                if (!languageMetaData["languageHash"]) {
+                    throw new Error("Could not find languageHash value inside languageMetaData object")
+                }
+                //Check the hash matches and then install!
+                if (languageHash == languageMetaData["languageHash"]) {
+                    //TODO: in here we are getting the source again even though we have already done that before, implement installLocalLanguage()?
```

Comment:
Hmm I thought we should confirm that the language source is really = `address` and languageMetaData["languageHash"], since perhaps somehow there is a vulnerability in the language language where you could attach a languageSource to any key without the hashing matching? Maybe I am too paranoid here and IPFS CID doesn't allow this to happen at all, lmk.

Reply:
Yeah, you're right. The meta data is the stuff that actually gets signed by the agent. So to really be cryptographically save against an attack we should do this.